packages/zoe/src/contractSupport/zoeHelpers.js

```diff
@@ -3,438 +3,363 @@
 import { assert, details } from '@agoric/assert';
 import { sameStructure } from '@agoric/same-structure';
 import { E } from '@agoric/eventual-send';
-import { satisfiesWant, isOfferSafe } from '../offerSafety';
+
+import { MathKind } from '@agoric/ertp';
+import { satisfiesWant } from '../contractFacet/offerSafety';
 
 import '../../exported';
 
 export const defaultRejectMsg = `The offer was invalid. Please check your refund.`;
 export const defaultAcceptanceMsg = `The offer has been accepted. Once the contract has been completed, please check your payout`;
 
-const getKeys = obj => harden(Object.getOwnPropertyNames(obj || {}));
 const getKeysSorted = obj =>
   harden(Object.getOwnPropertyNames(obj || {}).sort());
+
+// Compare actual keys to expected keys. If expectedKeys is
+// undefined, return true trivially.
+const checkKeys = (actual, expected) => {
+  if (expected === undefined) {
+    return true;
+  }
+  return sameStructure(getKeysSorted(actual), getKeysSorted(expected));
+};
+
 /**
- * Makes an object with helper functions useful to zoe contracts.
- *
+ * Given toGains (an AmountKeywordRecord), and allocations (a pair,
+ * 'to' and 'from', of AmountKeywordRecords), all the entries in
+ * toGains will be added to 'to'. If fromLosses is defined, all the
+ * entries in fromLosses are subtracted from 'from'. (If fromLosses
+ * is not defined, toGains is subtracted from 'from'.)
  * @param {ContractFacet} zcf
+ * @param {FromToAllocations} allocations - the 'to' and 'from'
+ * allocations
+ * @param {AmountKeywordRecord} toGains - what should be gained in
+ * the 'to' allocation
+ * @param {AmountKeywordRecord} [fromLosses=toGains] - what should be lost in
+ * the 'from' allocation. If not defined, fromLosses is equal to
+ * toGains. Note that the total amounts should always be equal; it
+ * is the keywords that might be different.
+ * @returns {FromToAllocations} allocations - new allocations
+ *
+ * @typedef FromToAllocations
+ * @property {AmountKeywordRecord} from
+ * @property {AmountKeywordRecord} to
```

Comment:
I like the intentional distinction between `Allocation` vs `AmountKeywordRecord`, where the latter may represent a change to an allocation, but a change to an allocation is not itself an `Allocation`. Using this consistently will make the code (and APIs and types and docs) easier to understand.

```suggestion
 * @property {Allocation} from
 * @property {Allocation} to
```